src/StringTools/InternableString.Simple.cs

```diff
@@ -200,29 +200,47 @@ public override unsafe string ToString()
         /// <returns>A stable hashcode of the string represented by this instance.</returns>
         public override int GetHashCode()
         {
-            int hashCode = 5381;
+            uint hash = (5381 << 16) + 5381;
+            bool isOddIndex = false;
 
             if (_firstString != null)
             {
                 foreach (char ch in _firstString)
                 {
-                    unchecked
-                    {
-                        hashCode = hashCode * 33 ^ ch;
-                    }
+                    hash = HashOneCharacter(hash, ch, isOddIndex);
+                    isOddIndex = !isOddIndex;
                 }
             }
             else if (_builder != null)
             {
                 for (int i = 0; i < _builder.Length; i++)
                 {
-                    unchecked
-                    {
-                        hashCode = hashCode * 33 ^ _builder[i];
-                    }
+                    hash = HashOneCharacter(hash, _builder[i], isOddIndex);
+                    isOddIndex = !isOddIndex;
                 }
             }
-            return hashCode;
+            return (int)hash;
+        }
+
+        /// <summary>
+        /// A helper to hash one character.
+        /// </summary>
+        /// <param name="hash">The running hash code.</param>
+        /// <param name="ch">The character to hash.</param>
+        /// <param name="isOddIndex">True if the index of the character in the string is odd.</param>
+        /// <returns></returns>
+        private static uint HashOneCharacter(uint hash, char ch, bool isOddIndex)
```

Comment:
Is this inlined? I'd imagine a method call for every character could add significant overhead

Reply:
Yes, I just checked, it's inlined by 3.5 JIT. It doesn't recognize the bit-rotate pattern, though, I'll delete the comment.

```
05010538 8bc6            mov     eax,esi
0501053a c1e005          shl     eax,5
0501053d 8bd6            mov     edx,esi
0501053f c1ea1b          shr     edx,1Bh
05010542 0bc2            or      eax,edx
05010544 03c6            add     eax,esi
05010546 33c1            xor     eax,ecx
```